### ISSUE TITLE
Query processor doesn't support Postgres enum columns

### ISSUE THREAD

**Issue by motti** [metabase/metabase#1623]
If a Postgres databases uses [enum types](http://www.postgresql.org/docs/9.1/static/datatype-enum.html) then those enum column cannot be queried using the query builder.

it returns "We couldn't understand your question".

⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by camsaul** [comment]
When used in a prepared statement you have to add a cast

```sql
CREATE TYPE bird_type AS ENUM ('pigeon', 'toucan');

[...]

SELECT * FROM [...] WHERE bird_type = 'pigeon' -- ok

**Comment by underyx** [comment]
@camsaul this is making a bunch of our (important, like `device_type`) columns unusable :(

Is it maybe possible that 0.20's backend could support this? Just in case the refactor you were waiting for has already been done.

**Comment by salsakran** [comment]
@underyx we're wrapping up 0.20 and releasing in a day or two. We've been "baking" the release candidate on our internal instances and bug fixing. 

Unfortunately, any changes that touch something as central our underlying type system can't be thrown in at this late a time.

**Comment by underyx** [comment]
@salsakran Oh, I know all that, don't get me wrong! I meant to ask if now with 0.20's release it'd be even possible to triage this issue 🙃

**Comment by underyx** [comment]
@salsakran But it's awesome that the metabase maintainers' first thought always seems to be

>This should be shipped to the user tomorrow

:D

**Comment by hai-ld** [comment]
Could someone kindly point me a workaround for this issue? It'd be deeply appreciated.

**Comment by underyx** [comment]
@hai-ld it's possible to create custom operators that are able to compare enums to Barbara, but this will break other queries if you compare varchars to literals in them without explicitly specifying their type.

**Comment by JoeSham** [comment]
@hai-ld I am currently dealing with this by having a view of the table I want to use in metabase and in that view I just change the type of enums to texts (ie. `(trip_type::text) as trip_type`). Same with columns of `text[]` type etc.

**Comment by kirkchris** [comment]
This would be awesome to get this working! Any updates on when it'll be going in?

**Comment by milgner** [comment]
An error message from the current version:

```
ERROR: operator does not exist: device_type_type = character varying Hint: No operator matches the given name and argument type(s). You might need to add explicit type casts. Position: 485
```

Strangely enough, when querying via SQL, the typecast works without problems.

**Comment by marlonbernardes** [comment]
+1

**Comment by marlonbernardes** [comment]
I'm having the same error, @milgner. Hopefully, this will be fixed soon.

**Comment by marlonbernardes** [comment]
Inside of `postgres.clj` (in `defn- prepare-value`), I managed to test a possible fix for this issue simply by returning a PGObject instead of directly returning the value:
 
```clojure
(doto (org.postgresql.util.PGobject.)
                                    (.setType "<enum name here>")
                                    (.setValue (name value)))
```

Unfortunately, I can't fix it yet, as I have roughly 2 hours of experience with Clojure and I'm still a bit confused with the syntax (shame on me, I know) :P

**Comment by mikeroelens** [comment]
Are there any plans for implementing a fix? The query builder is absolutely amazing, but because of this issue I can rarely use it and that makes me sad.

**Comment by JoeSham** [comment]
Any update on this? When is a fix for this planned?

**Comment by camsaul** [comment]
Created a test DB with data as such:

```sql
CREATE TYPE bird_type AS ENUM ('toucan', 'pigeon', 'parrot', 'finch', 'sparrow', 'turky');

CREATE TYPE bird_status AS ENUM ('good bird', 'bad bird', 'neutral', 'unknown', 'delicious');

CREATE TABLE birds (
  name varchar PRIMARY KEY NOT NULL,
  type bird_type NOT NULL,
  status bird_status NOT NULL
)

INSERT INTO birds ("name", "type", status) VALUES
  ('Rasta', 'toucan', 'good bird'),
  ('Lucky', 'pigeon', 'bad bird'),
  ('Freddy', 'finch', 'unknown'),
  ('Sammy', 'sparrow', neutral'),
  ('Theodore', 'turky', 'delicious')
```

What's interesting is a custom SQL query like this works without issue:

```sql
SELECT *
FROM birds
WHERE status = 'good bird'
```

(Postgres is smart enough to understand a string literal like `'good bird'` is meant as a `bird_status`)

But the equivalent query generated by our Query Processor:

```sql
SELECT *
FROM birds
WHERE status = ? -- ? is passed in as a String
```
doesn't work (something like "cannot compare type character varying with type bird_status").

So what we'd need to do is

```sql
SELECT *
FROM birds
WHERE status = ? -- ?::bird_status
```

which means we'll need to record this info somewhere so we can inject the appropriate casting info

**Comment by camsaul** [comment]
Alternatively since we're trying to compare strings anyway maybe it makes sense to just generate these sorts of `WHERE` clauses for Postgres like

```sql
SELECT *
FROM birds
WHERE status::text = ?
```

(always cast the thing that's going to be compared to a text type as `text`) which would likely fix a whole class of similar bugs.

**Comment by marlonbernardes** [comment]
@camsaul casting everything to text is not really feasible, as many enum type columns are indexed and casting it to text would make the indexes not to be applied (thus, resulting in slow queries).

**Setup:** (based on your scripts above)
```sql
CREATE INDEX bird_type_uk on birds(type)
```

**Example 1:** _explain select * from birds b where b.type = 'finch'_

```sql
-- Good, the index is being applied!
Bitmap Heap Scan on birds b  (cost=4.20..13.67 rows=6 width=40)
Recheck Cond: (type = 'finch'::bird_type)
->  Bitmap Index Scan on bird_type_uk  (cost=0.00..4.20 rows=6 width=0)
Index Cond: (type = 'finch'::bird_type)
```

**Example 2:** _explain select * from birds b where b.type = 'finch'_

```sql
-- Bad, this is performing a full table scan
Seq Scan on birds b  (cost=0.00..31.00 rows=6 width=40)
Filter: ((type)::text = 'finch'::text)
```

So, in short, one of the things that could be done is to use your first proposed solution, as it would result in the indexes being applied as well:

```sql
SELECT * FROM birds WHERE status = ?::bird_type
```

**Comment by camsaul** [comment]
Implemented by #6444

**Comment by winou75** [comment]
Hey guys, i still have the issue in latest version 0.27.2 where i am trying to filter by an attribute's value:

`ERROR: operator does not exist: "enum_Table_attribute" = character varying Hint: No operator matches the given name and argument type(s). You might need to add explicit type casts. Position: 1694`

is this fixed? or how can i change it locally?

**Comment by underyx** [comment]
@winou75 my assumption is that the fix is going to be included in 0.28 first, as it's a fairly large change.

**Comment by camsaul** [comment]
@winou75 @underyx that's right, we're shipping this as part of 0.28.

**Comment by JoeSham** [comment]
@camsaul does this fix also solve the issue with Field filter on an enum type column?

**Comment by jannehietamaki** [comment]
I'm still getting the error with v0.28.1.

`ERROR: operator does not exist: payment_type = character varying Hint: No operator matches the given name and argument type(s). You might need to add explicit type casts. Position: 271 `

**Comment by nfm** [comment]
@jannehietamaki I saw this too after upgrading from 0.27.2 to 0.28.1.

Re-syncing the database schema and clearing and re-scanning the cached field values fixed this for me. I can now use the question builder to write queries that filter on Postgres enum columns successfully.

**Comment by jannehietamaki** [comment]
@nfm That fixed it, thanks a lot!

**Comment by JoeSham** [comment]
@camsaul 

So I tried the query builder and it now works fine with the enums. But when I try to use the Field filter variable and map it to an enum column, it still doesn't work:
```
SELECT * FROM mytable
WHERE {{trip_type}};
```
This gives me
```
ERROR: operator does not exist: trip_type = character varying Hint: No operator matches the given name and argument type(s). You might need to add explicit type casts. Position: 193
```
:/

**Comment by FallenChromium** [comment]
Well, I'm not sure if this issue covers my problem, but I'm using PostgreSQL 10.5 and I can't assign aliases to enum fields in Metabase, there's only "show original values" option.

**Comment by jonasformolo** [comment]
I'm using Metabase v0.31.0 with PostgreSQL 9.4.20 and having this same issue when trying to filter by enum field in query builder:
Query failure: org.postgresql.util.PSQLException: ERROR: operator does not exist: basicos.enum_tp_pessoa = character varying Hint: No operator matches the given name and argument type(s). You might need to add explicit type casts. Position: 740
Is there any solution yet?

**Comment by jonasformolo** [comment]
I found a workaround for the issue above. When metabase reads the table definition, it sets the metabase_field.base_type as **type/*** and metabase_field.database_type as **"basicos"."enum_tp_pessoa"**. Updating these values to **type/PostgresEnum** and **basicos.enum_tp_pessoa** solves the issue. Maybe it could be fixed by someone experienced on Clojure. As far as I tested, this issue only happens when the enum type is defined on a specific schema ("basicos" in my case) and not on the public schema.

**Comment by sbelak** [comment]
I'm hitting the same issue on our ee.stats server. Field metadata seems correct (minus the initial misclassification of what `state` means).

**Comment by alichaudry** [comment]
I'm running into the same problem as @JoeSham -- it works fine as a native query, but when I try to do it as a `Field filter` and choose the filter from the dropdown it doesn't work. I get the following error:

```sql
ERROR: operator does not exist: recordstatus = character varying Hint: No operator matches the given name and argument type(s). You might need to add explicit type casts. Position: 3236
```

**Comment by flamber** [comment]
Shouldn't this be closed in favor of #7092, since the problem is specific to Field Filter?

**Comment by flamber** [comment]
Closing in favor of #7092
